vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,495 @@
+	StdinPipe() (io.WriteCloser, error)
+	StdoutPipe() (io.ReadCloser, error)
+	StderrPipe() (io.ReadCloser, error)
+	Start() error
+	Wait() error
+}
+
+func makeDefaultParams() *proxyAppParams {
+	return &proxyAppParams{
+		CommandRunner:   osutilCommand,
+		MakePipe:        osutil.LongPipe,
+		InitRetryDelay:  10 * time.Second,
+		MaxInitAttempts: math.MaxInt64,
+	}
+}
+
+func init() {
+	vmimpl.Register(
+		"proxyapp",
+		func(env *vmimpl.Env) (vmimpl.Pool, error) {
+			return ctor(makeDefaultParams(), env)
+		},
+		false)
+}
+
+type pool struct {
+	mu              sync.Mutex
+	isValid         bool
+	env             *vmimpl.Env
+	Proxy           *ProxyApp
+	ctxRunner       context.Context
+	ctxRunnerCancel context.CancelFunc
+	size            int
+}
+
+func (pool *pool) reset() {
+	pool.cancelRun()
+	pool.ctxRunner, pool.ctxRunnerCancel = context.WithCancel(context.TODO())
+	pool.Proxy = nil
+	pool.size = 0
+	pool.isValid = false
+}
+
+func (pool *pool) cancelRun() {
+	if pool.ctxRunnerCancel != nil {
+		pool.ctxRunnerCancel()
+		pool.ctxRunnerCancel = nil
+	}
+}
+
+func initPool(params *proxyAppParams, pool *pool, command string, pluginConfig []byte, env *vmimpl.Env,
+) error {
+	pool.reset()
+
+	var err error
+	pool.Proxy, err = runProxyApp(pool.ctxRunner, params, command)
+	if err != nil {
+		return fmt.Errorf("failed to run ProxyApp: %w", err)
+	}
+
+	pool.size, err = pool.Proxy.ConstructPool(string(pluginConfig), env.Debug)
+	if err != nil {
+		pool.cancelRun()
+		return fmt.Errorf("failed to construct pool: %w", err)
+	}
+
+	pool.isValid = true
+	return nil
+}
+
+type Config struct {
+	Command        string          `json:"proxy_cmd"`
+	ProxyAppConfig json.RawMessage `json:"proxyapp_config"`
+}
+
+func parseConfig(conf []byte) (*Config, error) {
+	vmCfg := new(Config)
+	if err := config.LoadData(conf, vmCfg); err != nil {
+		return nil, fmt.Errorf("failed to parseConfig(): %w", err)
+	}
+	return vmCfg, nil
+}
+
+// Package configuration VARs are mostly needed for tests.
+type proxyAppParams struct {
+	CommandRunner   func(bin string, args ...string) subProcessCmd
+	MakePipe        func() (io.ReadCloser, io.WriteCloser, error)
+	InitRetryDelay  time.Duration
+	MaxInitAttempts int64
+}
+
+func ctor(params *proxyAppParams, env *vmimpl.Env) (vmimpl.Pool, error) {
+	subConfig, err := parseConfig(env.Config)
+	if err != nil {
+		return nil, fmt.Errorf("config parse error: %w", err)
+	}
+
+	pool := &pool{
+		env: env,
+	}
+	pool.mu.Lock()
+	minWait := time.After(0)
+	initialized := make(chan bool)
+	attempt := int64(0)
+
+	go func() {
+		for {
+			attempt++
+			if attempt > params.MaxInitAttempts {
+				pool.mu.Unlock()
+				initialized <- false
+				return
+			}
+
+			<-minWait
+			minWait = time.After(params.InitRetryDelay)
+
+			err := initPool(params, pool, subConfig.Command, subConfig.ProxyAppConfig, env)
+			if err != nil {
+				log.Logf(0, "failed to initPool: %v", err)
+				continue
+			}
+
+			initialized <- true
```

Comment:
IIRC sending over a closed channel should lead to a panic. If indeed so, I wonder why it passed tests.

Reply:
I don't see code sending to the closed channel.

Reply:
Below you do 

```
initResult := <-initialized
close(initialized)
```

but the goroutine continues running and the next reconnection attempt it will again write to initialized. Or am I missing something?

Reply:
added todo for the stability test